tests/integration/conftest.py

```diff
@@ -344,18 +341,62 @@ async def app_fixture(
         "github-runner-image-builder/refs/heads/main/tests/integration/testdata/test_script.sh",
         SCRIPT_SECRET_CONFIG_NAME: "TEST_SECRET=TEST_VALUE",
     }
+
+
+@pytest.fixture(scope="module", name="base_machine_constraint")
+def base_machine_constraint_fixture(
+    private_endpoint_configs: PrivateEndpointConfigs, use_private_endpoint: bool
+) -> str:
+    """The base machine constraint."""
     num_cores = multiprocessing.cpu_count() - 1
     base_machine_constraint = f"arch={private_endpoint_configs['arch']} cores={num_cores} mem=16G"
     if use_private_endpoint:
         base_machine_constraint += " root-disk=100G"
     else:
         base_machine_constraint += " root-disk=80G"
+    return base_machine_constraint
+
+
+@pytest_asyncio.fixture(scope="module", name="app")
+async def app_fixture(
+    app_config: dict,
+    base_machine_constraint: str,
+    test_configs: TestConfigs,
+) -> AsyncGenerator[Application, None]:
+    """The deployed application fixture."""
     logger.info("Deploying image builder: %s", test_configs.dispatch_time)
     app: Application = await test_configs.model.deploy(
         test_configs.charm_file,
         application_name=f"image-builder-operator-{test_configs.test_id}",
         constraints=base_machine_constraint,
-        config=config,
+        config=app_config,
+    )
+    # This takes long due to having to wait for the machine to come up.
+    await test_configs.model.wait_for_idle(apps=[app.name], idle_period=30, timeout=60 * 30)
+
+    yield app
+
+    # Do not clean up due to Juju bug in model.remove_application. However, manual cleanup is
+    # required on private-endpoint OpenStack resources.
+    await test_configs.model.remove_application(app_name=app.name)
+
+
+@pytest_asyncio.fixture(scope="module", name="app_on_charmhub")
+async def app_on_charmhub_fixture(
+    test_configs: TestConfigs,
+    app_config: dict,
+    base_machine_constraint: str,
+) -> AsyncGenerator[Application, None]:
+    """Fixture for deploying the charm from charmhub."""
+    # Normally we would use latest/stable without pinning a revision here, but upgrading
+    # from stable is currently broken, and therefore we are using edge. Change this in the future.
```

Comment:
I decided to not add code complexity to allow upgrading from chroot to external mode without redeployment.